### PR TITLE
Display information when downloading a file.

### PR DESCRIPTION
**Pull request by FlorentTorregrosa**
Currently when downloading files, the command line displays a concatenated string of "Downloading: 100%" which is not helpful to see which file is downloaded where and which file fails downloading in case of failure.

Here is some code to improve the output of the command line.

Thanks for the review.

### CODE REVIEW

**Comment by T2L** [comment]
Also, it would be great if we can respect `--no-progress` composer option and skip showing this information at all.

**Comment by T2L** [comment]
I've checked the output. I my opinion it generates too much noise (and new lines). Suggestions: 

- combine into a single line (as composer does with vendor dependencies)
- add colors

**Comment by webflo** [comment]
Hi, i think the proposed changes by @T2L are valid. @FlorentTorregrosa do you want to change the PR accordingly?

**Comment by FlorentTorregrosa** [comment]
Hello,

Thanks both for your feedback. I will try to change the pull request accordingly. I also think that it would be nice to add colors and to do as composer does, I currently don't know how to do that. That's why I didn't do that before. I will search for that.

I don't know when I would have time to update my PR.

Also if I can have a feedback on https://github.com/drupal-composer/drupal-scaffold/pull/58 because all concerns where addressed, it would be nice so if something needs to be changed I will do it when I will (re-)dig into composer.

**Comment by FlorentTorregrosa** [comment]
Hello,

I have updated the pull request with the suggestions and fixed the tests.

Is it ok now to be merged?

File: src/PrestissimoFileFetcher.php [hunk 1]
```diff
@@ -43,7 +37,16 @@ protected function fetchWithPrestissimo($version, $destination) {
     array_walk($this->filenames, function ($filename) use ($version, $destination, &$requests) {
       $url = $this->getUri($filename, $version);
       $this->fs->ensureDirectoryExists($destination . '/' . dirname($filename));
-      $requests[] = new CopyRequest($url, $destination . '/' . $filename, false, $this->io, $this->config);
+      if ($this->progress) {
```
Comment (webflo) [inline src/PrestissimoFileFetcher.php]: I think, we can skip this hunk and add the information in line 65.

Comment (FlorentTorregrosa) [inline src/PrestissimoFileFetcher.php]: @webflo Thanks for your review.
At line 65, the $filename variable is not defined. So should I remove lines 40 to 49 and use the previous line 46? And leave line 65 as it is?

Comment (FlorentTorregrosa) [reply src/PrestissimoFileFetcher.php]: Seen with @greg-1-anderson. Will update the pull request.

**Comment by FlorentTorregrosa** [comment]
Saw with @greg-1-anderson, I will refactor FileFetcher, InitialFileFetcher and PrestissimoFileFetcher to avoid code duplication and to use Prestissimo for initial file fetching.

**Comment by clemens-tolboom** [comment]
I was about to submit a PR way worse then this one. 👍 

I learned __drupal-scaffold__ is fetching the d.o. files and it's good to know these are fetched and when refetching with `composer run-script 'drupal-scaffold'`.

**Comment by FlorentTorregrosa** [comment]
Hello,

The refactoring has been done. Thanks for the review.

File: src/FileFetcher.php [hunk 2]
```diff
@@ -17,23 +18,52 @@ class FileFetcher {
...
-  public function fetch($version, $destination) {
-    array_walk($this->filenames, function ($filename) use ($version, $destination) {
-      $url = $this->getUri($filename, $version);
-      $this->fs->ensureDirectoryExists($destination . '/' . dirname($filename));
-      $this->remoteFilesystem->copy($url, $url, $destination . '/' . $filename);
-    });
+  public function fetch($version, $destination, $erase) {
+    foreach ($this->filenames as $sourceFilename => $filename) {
+      $target = "$destination/$filename";
+      if ($erase || !file_exists($target)) {
```
[the remark this answers is not in the document]
Comment (FlorentTorregrosa) [reply src/FileFetcher.php]: It fetches a file even if it already exists be cause in https://github.com/drupal-composer/drupal-scaffold/pull/65/files#diff-bb433a78829fe2e46d711575194d4f50R140 we use the $erase parameter to TRUE and only $erase to FALSE when using it for the initial fetch https://github.com/drupal-composer/drupal-scaffold/pull/65/files#diff-bb433a78829fe2e46d711575194d4f50R143

In the README.md:
The initial hash lists files that should be copied over only if they do not exist in the destination. The key specifies the path to the source file, and the value indicates the path to the destination file.

[the remark this answers is not in the document]
Comment (clemens-tolboom) [reply src/FileFetcher.php]: I was referring @ the code __lines__ changed. I should have dug deeper. _Sorry_

[the remark this answers is not in the document]
Comment (FlorentTorregrosa) [reply src/FileFetcher.php]: No problem. Thanks for the review.

**Comment by FlorentTorregrosa** [comment]
Hello,

I have rebased my changes to followup the last code changes in drupal-scaffold.

Would it be possible to have a review please?

File: src/FileFetcher.php [hunk 3]
```diff
@@ -32,7 +39,13 @@ public function fetch($version, $destination) {
     array_walk($this->filenames, function ($filename) use ($version, $destination) {
       $url = $this->getUri($filename, $version);
       $this->fs->ensureDirectoryExists($destination . '/' . dirname($filename));
+      $this->io->write("Going to download the file $filename");
+      $this->io->write("  from: $url");
+      $this->io->write("  to: $destination/$filename");
```
Comment (clemens-tolboom) [inline src/FileFetcher.php]: You better change this into a single line
```
$this->io->write("Fetching $url into $destination/$filename");
```
That makes the output shorter in lines and allows to grep for ie .htaccess and get the full line.

Comment (FlorentTorregrosa) [reply src/FileFetcher.php]: @clemens-tolboom, thanks for the review, but the comment is not on the last codebase. Those lines had changed in https://github.com/drupal-composer/drupal-scaffold/pull/65/commits/e424668d2360958b52417f72a4d9824b9ff170c6.

**Comment by webflo** [comment]
The ``$erase`` parameter is new feature? It was don't there before.

**Comment by FlorentTorregrosa** [comment]
Hello,

The $erase is not a new feature, it was introduced to refactor the FileFetcher and the initialFileFetcher classes. Therefore the initial fetch also benefits from the PrestissimoFileFetcher.

File: src/FileFetcher.php (hunk 3 above)
Comment (clemens-tolboom) [reply src/FileFetcher.php]: Hmmm ... guess I'm commenting through the wrong page. (darn) ... let's delete this crap

**Comment by webflo** [comment]
Alright, i understand it has been a while since i looked at the code. Lets rename it to ``$override``? I thinks this is a better fit.

**Comment by FlorentTorregrosa** [comment]
OK. I have changed the name of the variable.

**Comment by webflo** [comment]
Merged. 

@FlorentTorregrosa thanks 👍

**Comment by FlorentTorregrosa** [comment]
\o/ Thanks!!! Finally merged :gift: :christmas_tree: :gift:. It's Christmas magic :D

Next steps:
* #58
* #64

**Pull request merged**